proposals/0000-complete-set-signatures.rst

```diff
@@ -0,0 +1,483 @@
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-} -- (1)
+
+ instance IsList [] where ...
+
+ f :: IsList l => l Int -> Int
+ f Empty    = 0
+ f (x :< _) = x
+
+ class IsList l => IsInfinite l where
+   -- INVARIANT: `viewCons` always returns `Just`
+
+ data Stream a = S a (Stream a)
+ instance IsList Stream where
+   viewCons (S x xs) = Just (x, xs)
+ instance IsInfinite Stream where
+
+ {-# COMPLETE (:<) :: IsInfinite l => l a #-} -- (2)
+
+ instance IsInfinite Stream where ...
+
+ safeHead :: IsInfinite l => l a -> a
+ safeHead (x :< _) = x
+
+ safeHead2 :: Stream a -> a
+ safeHead2 (x :< _) = x
+
+ unsafeHead :: [a] -> a
+ unsafeHead (x :< _) = x
+
+This program passes type-checking. The compiler *should* emit a warning about
+the definition of ``unsafeHead`` being incomplete, but not for ``f``,
+``safeHead`` or ``safeHead2``:
+
+- ``f`` has a case for ``Empty`` and ``(:<)``. COMPLETE set (1) is covered,
+  because the type of the pattern match is ``IsList l => l a``, which is
+  subsumed by itself. Thus, the pattern match of ``f`` is exhaustive.
+- ``f`` has a case for ``(:<)``. COMPLETE set (2) is *not* covered,
+  because the type of the pattern match is ``IsList l => l a``, which is not
+  subsumed by ``IsInfinite l => l a``.
+- ``safeHead`` has a case for ``(:<)``. COMPLETE set (2) is covered,
+  because the type of the pattern match is ``IsInfinite l => l a``, which is
+  subsumed by itself. Thus, the pattern match of ``safeHead`` is exhaustive.
+- ``safeHead2`` has a case for ``(:<)``. COMPLETE set (2) is covered,
+  because the type of the pattern match is ``Stream a``, which is
+  subsumed by ``IsInfinite l => l a``.
+  Thus, the pattern match of ``safeHead2`` is exhaustive.
+- ``unsafeHead`` has a case for ``(:<)``. COMPLETE set (2) is *not* covered,
+  because the type of the pattern match is ``[a]``, which is not
+  subsumed by ``IsInfinite l => l a``.
+- The lack of any COMPLETE set being covered by the the pattern match in
+  ``unsafeHead`` means that its definition is flagged as inexhaustive.
+
+Constraints in the signature
+============================
+
+Given the following example (full example in `#14422 <https://gitlab.haskell.org/ghc/ghc/-/issues/14422#note_296728>`_) ::
+
+ pattern FZ :: () =>        n ~ 'S m      => Fin n
+ pattern FZ <- (viewFin -> VZ) where
+   FZ = Fin 0
+
+ pattern FS :: () => n ~ 'S m => Fin m -> Fin n
+ pattern FS m <- (viewFin -> VS m) where
+   FS (Fin m) = Fin (1 + m)
+
+ {-# COMPLETE FZ, FS #-}
+
+ inc :: Fin n -> Fin (S n)
+ inc FZ       = FS FZ
+ inc n@(FS _) = FS n
+
+How would I give the signature a COMPLETE set while preserving its semantics?
+
+- ``{-# COMPLETE FZ, FS :: Fin n #-}`` is the obvious and correct choice.
+- ``{-# COMPLETE FZ :: Fin Z #-}; {-# COMPLETE FS :: Fin (S n) #-}`` Means that
+  neither COMPLETE set applies at a match type of ``Fin n``, as long as there is
+  no Given constraint that refines ``n``. The pattern-match in ``inc`` would be
+  regarded as incomplete.
```

Comment:
Well `{-# COMPLETE FZ :: Fin Z #-}` is still *correct* in the sense that if you match against a value of type `Fin Z` then only `FZ` can happen. So it may be correct, but it only applies in limited situations.

Moreover (and you should say this) the more general one `{-# COMPLETE FZ, FS :: Fin n #-}` completely subsumes it, using "subsumes" in the informal English-language sense.   If you *do* have a match on `Fin Z`, the GADT stuff means that we effectively thin down the FZ,FS to just FZ.  So the FZ,FS pragma does the job of both the FZ one and the FS one.

Ideally we'd compare COMPLETE pragmas and warn that the FZ one is redundant.

But regardless you should say something about GADTs and how they interact here.

Reply:
I think there is some confusion. Actually, I can do better: there is some confusion here. I say this because `Fin Z` is uninhabited, and `FZ` definitely inhabits `Fin (S n)` for all `n`.

Let's say, though, that we have

```
{-# COMPLETE K1 :: T Z #-}
{-# COMPLETE K1, K2 :: T n #-}
```

I'd say neither set subsumes the other. In the case where `n ~ Z`, the first one subsumes the second -- not the other way around! That is, a pattern match at type `T Z` containing both `K1` and `K2` would be complete, as we could deduce just from the first `COMPLETE` pragma above. (It is more than complete: it has a redundant pattern -- provided `K2` comes after `K1`.)